### PR TITLE
Input control validation on focus out event

### LINKED ISSUE #2023: [🐛 BUG] [UX] input, number, date controls don't validate on focus out event
### What went wrong? 🤔

The input control (as well as number, data, ...) does not trap the Focus Out event that should trigger the equivalent of pressing the Enter key

### Expected Behavior

See description.

All `input`, `number`, `date` should be impacted.
`login` probably...

image

In this case, I don't want to be pressing Enter before "Share".

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by kart2004**
Description:
- Fixes #2023 
- Implemented onBlur event handler in the Input.tsx component to trigger value updates when an input field loses focus (focus out event). 
- Ensures that changes are captured even when the Enter key is not pressed, improving the overall user experience. 
- The change impacts text, number, and date input fields to align with the intended behavior.

image

Key changes include:
- Added onBlur handler to TextField for focus out validation.
- Ensured that value updates propagate correctly upon focus loss.

Ensured 
- Code is rebased 
- License is present 
- pre-commit works - without mypy error.
- Taipy tests are passing.
- 
image
image

### CODE REVIEW

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 1]
```diff
@@ -138,6 +138,20 @@ const Input = (props: TaipyInputProps) => {
         [changeDelay, dispatch, updateVarName, module, onChange, propagate]
     );
 
+    const handleBlur = useCallback(
+        (e: React.FocusEvent<HTMLInputElement>) => {
+            const val = e.target.value;
+            if (changeDelay > 0 && delayCall.current > 0) {
+                clearTimeout(delayCall.current);
+                delayCall.current = -1;
+                dispatch(createSendUpdateAction(updateVarName, val, module, onChange, propagate));
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx]: ~~this should always be called~~
~~So put outside the if/else block~~
I think you're right this should only be called if there is delayed call on hold
Sorry about the noise

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
I don't see any tests

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 1 above)
Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx]: But I still don't see any tests ;-)
Are you still working on this @kart2004  ?

**Comment by RymMichaut** [comment]
hey @kart2004, congrats on the merged PR.
Could you please reach out at rym.michaut@taipy.io to organize your shipment?

**Comment by FabienLelaquais** [comment]
@kart2004
As you certainly already know, the team has granted you the Hacktoberfest reward for your job on this matter.
Tests are still missing (as pointed out by @FredLL-Avaiga ), and I want to slightly change the original requirements after discussing that matter with actual users: we need to add a new Boolean property (like `validate_on_blur`) that if True would trigger the equivalent of pressing the Enter key should the *change_delay* property be set to 0 (no immediate update of the backend data).

Please let us know if you wish to finish this work or let us handle this.
Thanks!

**Comment by RymMichaut** [comment]
please @kart2004 respond before EoW to ship, or you'd miss your swag

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 1 above)
Comment (FabienLelaquais) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx]: Tests were added.

**Comment by FabienLelaquais** [comment]
Finally, the new property is called 'action_on_blur' for homogeneity.

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 2]
```diff
@@ -138,6 +139,27 @@ const Input = (props: TaipyInputProps) => {
         [changeDelay, dispatch, updateVarName, module, onChange, propagate]
     );
 
+    const handleBlur = useCallback(
+        (evt: React.FocusEvent<HTMLInputElement>) => {
+            let val = (type === "number")
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx]: should be a `const`

File: frontend/taipy-gui/src/components/Taipy/Input.tsx [hunk 3]
```diff
@@ -138,6 +139,27 @@ const Input = (props: TaipyInputProps) => {
...
+                    : evt.currentTarget.querySelector("input")?.value)
+                ;
+            if (delayCall.current > 0) {
+                if (changeDelay > 0) {
+                    clearTimeout(delayCall.current);
+                    delayCall.current = -1;
+                }
+                dispatch(createSendUpdateAction(updateVarName, val, module, onChange, propagate));
+            }
+            onAction && dispatch(createSendActionNameAction(id, module, onAction, "Tab", updateVarName, val));
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Input.tsx:157]: "Tab" => "Blur"

File: taipy/gui/viselements.json [hunk 4]
```diff
@@ -114,31 +114,37 @@
...
-                        "doc": "The type of generated input HTML element, as defined in [HTML input types](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input_types).<br/>This value forces certain values to be entered and can be set to \"text\", \"tel\", \"email\", \"url\"..., among other choices."
+                        "doc": "The type of input element, as per <a href=\"https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input_types\">HTML input types</a>.<br/>This property enforces specific input formats where applicable. Supported values include \"text\", \"tel\", \"email\", \"url\", etc."
+                    },
+                    {
+                        "name": "action_on_blur",
+                        "type": "bool",
+                        "default_value": "False",
+                        "doc": "If True, the <code>on_action</code> callback is triggered when the input control looses keyboard focus (e.g., when the user presses the Tab key). When this happens, the key name for the event (set in the <i>args</i> property of the <i>payload</i> parameter to the callback function) is set to \"Tab\"."
```
Comment (FredLL-Avaiga) [inline taipy/gui/viselements.json:147]: "Tab" => "Blur" ?

File: taipy/gui/viselements.json [hunk 5]
```diff
@@ -162,29 +168,35 @@
...
                         "name": "max",
                         "type": "dynamic(Union[int,float])",
-                        "doc": "The maximum value to accept for this input."
+                        "doc": "The maximum acceptable value.<br/>Values above this threshold are invalid."
+                    },
+                    {
+                        "name": "action_on_blur",
+                        "type": "bool",
+                        "default_value": "False",
+                        "doc": "If True, the <code>on_action</code> callback is triggered when the number control looses keyboard focus (e.g., when the user presses the Tab key). When this happens, the key name for the event (set in the <i>args</i> property of the <i>payload</i> parameter to the callback function) is set to \"Tab\"."
```
Comment (FredLL-Avaiga) [inline taipy/gui/viselements.json:199]: "Tab"

**Review by FredLL-Avaiga — DISMISSED** [review]

File: taipy/gui/viselements.json (hunk 4 above)
Comment (FabienLelaquais) [reply taipy/gui/viselements.json:147]: Not a key name... let's talk about this.

File: taipy/gui/viselements.json (hunk 5 above)
Comment (FabienLelaquais) [reply taipy/gui/viselements.json:199]: That depends on the context.
In the whole document, keys have no quotes.
Strings do.

File: frontend/taipy-gui/src/components/Taipy/Input.tsx (hunk 2 above)
Comment (FabienLelaquais) [reply frontend/taipy-gui/src/components/Taipy/Input.tsx]: MUST be, yes!

**Review by FredLL-Avaiga — APPROVED** [review]

**Comment by FabienLelaquais** [comment]
@kart2004 
FYI - Your PR was slightly modified to, in our opinion, better match the requirements that we've had.
Thanks again for your effort.

**Pull request merged**